src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 and BSD-3-Clause
+
+//! System device management.
+//!
+//! [DeviceManager](struct.DeviceManager.html) responds to manage all devices
+//! of virtual machine, store basic device information like name and
+//! parent bus, register IO resources callback, unregister devices and help
+//! VM IO exit handling.
+
+extern crate vm_allocator;
+
+use self::vm_allocator::{Error as AllocatorError, SystemAllocator};
+use crate::device::*;
+use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
+use std::collections::btree_map::BTreeMap;
+use std::collections::HashMap;
+use std::result;
+use std::sync::{Arc, Mutex};
+use vm_memory::{Address, GuestAddress, GuestUsize};
+
+/// Guest physical address and size pair to describe a range.
+#[derive(Eq, Debug, Copy, Clone)]
+pub struct Range(pub GuestAddress, pub GuestUsize);
+
+impl PartialEq for Range {
+    fn eq(&self, other: &Range) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Ord for Range {
+    fn cmp(&self, other: &Range) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl PartialOrd for Range {
+    fn partial_cmp(&self, other: &Range) -> Option<Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+/// Error type for `DeviceManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The insertion failed because the new device overlapped with an old device.
+    Overlap,
+    /// PIO request is none.
+    NonePIORequest,
+    /// The insertion failed because device already exists.
+    Exist,
+    /// The removing fails because the device doesn't exist.
+    NonExist,
+    /// Irq allocation failed.
+    IrqAllocate(AllocatorError),
+    /// Instance id allocation failed.
+    InstanceIdAllocate(AllocatorError),
+    /// Address allocation failed.
+    AddressAllocate(AllocatorError),
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System device manager serving for all devices management and VM exit handling.
+pub struct DeviceManager {
+    /// System allocator reference.
+    resource: Arc<Mutex<SystemAllocator>>,
```

Comment:
All fields in SystemAllocator are already protected by Arc<Mutex<>>, why another level of Arc<Mutex<>>?
Seems we need to specify who takes the responsibility to protect DeviceManager instance from concurrent access. Suggestion the owner of the DeviceManager object takes the responsibility to protect from concurrent access. If the vmm, such as firecracker,  supports static device manager, there's need for Mutex at all. On the other hand, VMM like qemu may protect the DeviceManager object withj Arc<Mutex<>>.

Reply:
Yes, I have a new patch removing the protection here, since we want SystemAllocator to be thread safe of each members, and also, vmm would have a static DeviceManager being protected.